pkg/kubectl/cmd/config/config.go

```diff
@@ -66,59 +80,323 @@ func NewCmdConfig(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+		}
+
+	}
+
+	return nil
+}
+
+func (o *PathOptions) getStartingConfig() (*clientcmdapi.Config, error) {
+	if err := o.Validate(); err != nil {
+		return nil, err
+	}
+
 	config := clientcmdapi.NewConfig()
 
-	if len(o.specifiedFile) > 0 {
-		filename = o.specifiedFile
-		config = getConfigFromFileOrDie(filename)
+	switch {
+	case o.Global:
+		config = getConfigFromFileOrDie(o.GlobalFile)
+
+	case o.UseEnvVar:
+		config = getConfigFromFileOrDie(o.EnvVarFile)
+
+	case o.Local:
+		config = getConfigFromFileOrDie(o.LocalFile)
+
+		// no specific flag was set, load according to the loading rules
+	default:
+		clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(o.LoadingRules, &clientcmd.ConfigOverrides{})
+		rawConfig, err := clientConfig.RawConfig()
+		if err != nil {
+			return nil, err
+		}
+		config = &rawConfig
+
 	}
 
-	if o.global {
-		if len(filename) > 0 {
-			return nil, "", fmt.Errorf("already loading from %v, cannot specify global as well", filename)
+	return config, nil
+}
+
+// GetDefaultFilename returns the name of the file you should write into (create if necessary), if you're trying to create
+// a new stanza as opposed to updating an existing one.
+func (o *PathOptions) GetDefaultFilename() string {
+	switch {
+	case len(o.LoadingRules.ExplicitPath) > 0:
+		return o.LoadingRules.ExplicitPath
+
+	case o.Global:
+		return o.GlobalFile
+
+	case o.UseEnvVar:
+		return o.EnvVarFile
+
+	case o.Local:
+		return o.LocalFile
+
+		// if there's a local file, use that; else if there's an envvar, use that; else use the global
+	default:
+		if _, err := os.Stat(o.LocalFile); err == nil {
+			return o.LocalFile
 		}
 
-		filename = os.Getenv("HOME") + "/.kube/.kubeconfig"
-		config = getConfigFromFileOrDie(filename)
+		if len(o.EnvVarFile) > 0 {
+			return o.EnvVarFile
+		}
+
+		return o.GlobalFile
 	}
 
-	if o.envvar {
-		if len(filename) > 0 {
-			return nil, "", fmt.Errorf("already loading from %v, cannot specify global as well", filename)
+}
+
+// ModifyConfig takes a Config object, iterates through Clusters, AuthInfos, and Contexts, uses the LocationOfOrigin if specified or
+// uses the default destination file to write the results into.  This results in multiple file reads, but it's very easy to follow.
+// Preferences and CurrentContext should always be set in the default destination file.  Since we can't distinguish between empty and missing values
+// (no nil strings), we're forced have separate handling for them
+func (o *PathOptions) ModifyConfig(newConfig clientcmdapi.Config) error {
+	startingConfig, err := o.getStartingConfig()
+	if err != nil {
+		return err
+	}
+
+	// at this point, config and startingConfig should have, at most, one difference.  We need to chase the difference until we find it
+	// then we'll build a partial config object to call write upon.  Special case the test for current context and preferences since those
+	// always write to the default file.
```

Comment:
May want to mention somewhere that while it looks like this code will generate multiple file writes, in practice, a command only updates/removes one entry, and thus only writes to one file.